include/up-cpp/utils/Expected.h

```diff
@@ -22,87 +23,161 @@ namespace uprotocol::utils {
+	constexpr explicit Unexpected(const E& rhs) : storage(rhs) {}
+	constexpr explicit Unexpected(E&& rhs) : storage(std::move(rhs)) {}
+
+	template <class... Args>
+	constexpr explicit Unexpected(std::in_place_t, Args&&... args)
+	    : storage(args...) {}
+
+	// template <class U, class... Args>
+	// constexpr explicit Unexpected(std::in_place_t, std::initializer_list<U>
+	// il, Args&&... args)
+
+	constexpr const E& error() const& noexcept { return std::move(storage); }
+
+	// this one is checked
+	constexpr E& error() & noexcept { return storage; }
+
+	constexpr const E&& error() const&& noexcept { return std::move(storage); }
+
+	constexpr E&& error() && noexcept { return std::move(storage); }
+};
 
 /// @brief A stripped-down version of std::expected from C++23.
-template <typename T, typename E>
+template <typename E, typename U>
 struct Expected {
-	using value_type = T;
-	using error_type = E;
-
-	constexpr Expected();
-	constexpr Expected(const Expected& other);
-	constexpr Expected(Expected&& other) noexcept;
-
-	constexpr explicit Expected(T&& v);
-	constexpr explicit Expected(E&& e);
-
-	Expected& operator=(const Expected& other);
-	Expected& operator=(Expected&& other);
-
-	Expected& operator=(const T& other);
-	Expected& operator=(T&& other);
-
-	Expected& operator=(const E& other);
-	Expected& operator=(E&& other);
-
-	constexpr explicit operator bool() const noexcept;
-	constexpr bool has_value() const noexcept;
-
-	constexpr T& value() &;
-	constexpr const T& value() const&;
-	constexpr T&& value() &&;
-	constexpr const T&& value() const&&;
-
-	constexpr const E& error() const& noexcept;
-	constexpr E& error() & noexcept;
-	constexpr const E&& error() const&& noexcept;
-	constexpr E&& error() && noexcept;
-
-	constexpr T value_or(T&& default_value) const&;
-	constexpr T value_or(T&& default_value) &&;
-
-	template <typename F>
-	constexpr auto and_then(F&& f) &;
-	template <typename F>
-	constexpr auto and_then(F&& f) const&;
-	template <typename F>
-	constexpr auto and_then(F&& f) &&;
-	template <typename F>
-	constexpr auto and_then(F&& f) const&&;
-
-	template <typename F>
-	constexpr auto or_else(F&& f) &;
-	template <typename F>
-	constexpr auto or_else(F&& f) const&;
-	template <typename F>
-	constexpr auto or_else(F&& f) &&;
-	template <typename F>
-	constexpr auto or_else(F&& f) const&&;
+	std::variant<E, Unexpected<U>> storage;
+
+public:
+	constexpr Expected() { storage = E(); }
+	constexpr Expected(E&& rhs) : storage(std::move(rhs)) {}
+	constexpr Expected(Unexpected<U>&& rhs) : storage(std::move(rhs)) {}
+
+	constexpr Expected(const Unexpected<U>& rhs) {
+		std::get<Unexpected<U>>(storage).storage = std::move(rhs);
+	}
+
+	template <class X = E>
+	constexpr explicit Expected(X&& rhs) {
+		std::get<E>(storage) = std::forward<X>(rhs);
+	}
+
+	template <class X = U>
+	constexpr explicit Expected(Unexpected<X>&& rhs) {
+		std::get<E>(storage) = rhs.error();
+	}
```

Comment:
@debruce - You mentioned that this did not work. I tried it and it does work. However, there is a bug in your gtest code. You are using `auto sample = get_rand()` where the signature is `uint64_t get_rand();`, then trying to implicitly return `sample` as an `Expected<int, int>`. Once the test code is updated to make `sample` an `int`, the tests work.

[Compiler explorer](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGEgKykrgAyeAyYAHI%2BAEaYxCAAHACcpAAOqAqETgwe3r4BaRlZAqHhUSyx8cm2mPaOAkIETMQEuT5%2BXIF2mA7Zjc0EpZExcYkpCk0tbfmdtpODYcMVo8kAlLaoXsTI7BwA9HsA1ELKACIAGgC0AGJ0mB6pAJ7EeMAIBAAqmKoEIIdmGjMkkOHkEL2iXiIxAUhyIsIQmEO2GQtDwqSUh2umwY6CY9QYJg0AEEDoSSUchJhEQQEYcIgB5D4ASTc2EONHoEAUq0O%2BAm4MhmHQhwA7oQEPC8DCRSQANbskiHJjofD4sRkg6HML8YgsPHZQ7ETDAZr4IyHNBPF5vAiHVAi8LQhBogB0Gr27sOH2dMNSxFQwGIrCVOPhiKYyDQLFShkeYWAhz1BDieDEMOaiL1WCVADcmHQmNF6IcvDi4mHPcndTDUFQw4ciTHkLTQjsGBiAGpxTICf4ujSi53NrXpvMFovhgie94EdEgA4ixcuphNhEukjAPaottKBR7YIs7ARITYS5mfvuz0nC6XVuMJSXJlYQR4GijBurzBni/EsnmADMr5lnWcjKAA%2Bm4yjgXIzLBEIYHYOcyjYG4HzYKcYEABJ/mY/5YDQ4SHKBEFQWBMFMnBCFIShaEYdhv6/rhYQol42YmP%2BbgTFgqg7KkU7/tgOGAQwLFsRxBCPKkmBgQQQaEAo7GCYxwmiYi7FuHmLyGPxSlEmSzBsAoTaIl4fqoEQaC0CAICQnQMImAA7FYDHkg2hxKC%2B4S0IcspUqk8ainiw46oOjCioiChigQw5wm41jWLhIbCiaYRKpGmAKDCRCVrSRpiO5BDoNZ3xSQ4Qouoc9JbIcqIsIQQoWqgTzuQgmy0MKsSGpgqQGDswp4CwbD4HitSPG6xITPqyBgUwmVxAQEBgGAYFgQgs1gcgqSpDNBCyXgELJhAK2bdtqLRGBJU9Mm6DrGShz3Q9j2PeYZgAErdb1JlmRZnjWbZtAKNZ2CqKV12iuKBVFSAl1legL2rOxzl6cSBxHAAAgZiJfNGJDNI87leNEEyEIKCrEJDxUg1dQruujRp6sQsowpS1K5ZgVBxIwOxkxT0NU7DIC009T0znOByMC6J1GhzRoiZgkuoCwewinsJ17P9hCPHsMPXULwsPREqDslsNLlugqDID4jBNPiI6HGZOZ4Fg/UMJKMIcvLQto45SP8l4DiHBYyrA6DQpEulmWHH8qSE9uvPEKWjhsBdxD%2BuTPt/hoSRB%2BgIfU%2Bg4c7JlEA8n8XHWQnL7J3EacQC9wCMCmyCB8H/PXYXGUKWYZg8hnDmnJnyNJMm0YGMmhzqRJUmYy6s8NsQwAKQJmfZ63ocFxHCgQESC9dwAbOYe%2BzxVzSLwjDFJI9ZeFRXicDdJNckFyN8gDqIqmupO%2BL4pECn9yx%2B9yco5Aev5%2B6I0HqjQ4aNogvHZocd6ABHLweAjTCiaMAYAAUp6IhCqIXcoUjSHB1g1QwwpSzELQZJTAQsMxamfI4UQtBxpEhHj1EaE9xJUMxkiRSZIUSzRhHIcIbcGoZ2JDHIseBkCCwvmgdsyYQbkyERQiAciJhEWEevQ%2Bvd/ynF5OzJgXhaD8SRoSJIaiFF%2Bg0So5RIi4ZmAPg4nRej8KGOMeAlyZiLElXJiVbchBrF2NUQIdRgkHGGgQNyKOBVcYNwgMQSJgCrBgNkSEyxviQb%2BNtLY9eEAwmOL3hEqJ19YmYGflDFgqAcxlISdyJJwDB5mNYWPNSHF%2BGZWPvPb%2By9UnyJ8UQzJUiAk5PzuU6yYQwJsJ2DJUgXT96H06X/c%2BQ8r4xKDHE8ur8SDv2IHDDiX8l7YF/rvVYACJ5AJScjIe3jFGNXkTw8Jj9iAlzuRMQ%2BhwGCoG%2BLxW0PsuoEC2K7CYpTEbnJAVcrxaT%2BkWIeQUohqcn48gsYfd5nzvndV%2BU5f5gK1lMAbqChpnis43Ksfk95TyXkovCWiniGLzkWGxcQIFL9KnVOfqUhG/4GWEquX6PAeZkwyKHmyYF6zqFcr/GAiVLlIHQNgXWIkBUXhbSFJcc2DpDjVOhAaWsvMKHsn9CwEE8VLC4WYc09hk8uGsCxrM7B3DBI9PDgYSOedYb0rJBI7cQqAKvkOBoFeJLyZuuuhAD4VLCm1NLrijZLKqk1MSfUy5kK%2Bm3JDUKCAIzYbqUdWS8JUbomirxWUzZrKE11I9cm3pEx%2BnpvQME%2B5WbdYcVze8gtYih5mM2Q3fibgm00xbbwgS7KxWnKLQ3DhejS3xviYkjxnaq1DwtePdS7SYTnEnV6Xh1b0kDJ6kM20daIDnAjUU%2BpF8u0vx7epcNw7x2YGcbzN%2BH8OInuHVG%2BdZieUr2Xa0twa7DgbvYno1tO7%2Bl%2BIPUiIJ/a9luDfXmyNiSPUXqzt2zAvbc0juLY%2B2pLoKWctMVnb9YG01BJhXWttSGSlitnQoF096k3goArUJQd0Hq/qtdPG1nSDnbsVQ9INUHcm8YcQsueSzC1QmLWMrZxAdmwd48OpZZy%2B5Mdwq4V8jTiVQtudEVAnhDhrQUGBPM3gylIrSR8r5tK%2BLIc7VnI0AKmW81au1YzYgqzMEcNUm9P8GOfqI4ulNNbbkQeQAEpqcQ8SKj054F5MKaU/Ls1%2Bhz6GcWbNc%2Bgdzxi4hef5eK1CfmpMNwIyvYjQ9NS/o4f%2Bl166%2BNmM1IJj4mqxBeGkk/E9onwk5gs/I1F1mksdoa0cMxjmcVGZM618zHDri82vRxW9RyGNRxXpqSaDCNqzQw0VqGz7dnqXgxAHrpWL6anK00zAo9LVtNq4B%2Br2nU1WOa6ZtrYEOunp668qc1KBt0qG0R1LTnXYTZe9N9is20O9sW1hkrkndvbJfXBn%2BPWAtJHOw9kLViYXhu61N%2BLaT3n/bR1nP1i0wAg7x6sZZKXifDwQP6EULdc52I7sXMrAPL4PReiw1hto4QRiLjCUHLyRQIlduQuxbpu4nfs0kMbznIe%2BbvcVh9qP0fmJ009lrZnKVdcKUTsxpOloU511T9ntOaQM6Z3W1nW9zdmOet3Hnl3bP883trtrIuxclk0fnKXPc1eA/S3GtliuFtFY5TLr9QWMe7vI%2BShFzzetvPCQbkndYIAm892blDFv6f2mtyzzeddc8O65073aLu%2BdGwF53eFtceSi7Cp820Ev17%2B6jxz%2BXzKobzb7b77Ng7l7K45XhxPJc1cx41494N9fEUTz18ljnpOs/mepxzpp%2BfGc5xt8X%2B3WdHdmGd9GavaVBdz6T6FV2LefcUI74HuXaWFch5LVe9D6kYM5qHUtlXVOx8N8nyYxIyx0sxx0KUiyDChAACp8d7k08kgjdyd1phcc9ZdN8rcd8i9BcS80CD9y8j9K8T9YQjYsBpZOY5Y9184PdEQm9xcB9rp79pVZdu85t39w8R9R1ACtNp9MdyZscoC7QpJICSBLgf5k9MVCNadEDV8S519ad0CC9MD15bccCacy97pudCDXcSDOYZYuZER9V0gwgqwr9b9Jd4YH8WDD4w9CsODsMuCXI%2BUBV2AV5NlNJUxBAb1ZlP8h9c0Y1xVCMh51spEZo5oWgydNkpQTNUAnYTMldsBSAV59Z9YXoAB1REc2BgMADgW0MQWgAvR1AeXRHMGI4UCAUsVEXyPVOxBGaXALIIm2EIgReaVQvA%2B6SI4zLAP2XoCcOI9gtkU9JaDomaVOJgR4Po2wgYvXJI/WIYl%2BKIsg2WaZHMeIxIi%2BQ/OtLqJBFBDKJEYgxMKkPnWkOKSwawQ4bIro2Sf2JoCcbIrY5Bema2CwpghpF4lGTUb2ILfue6TUTGIyCML6f0H6KyGyRwAGRpdTHEP1Q4TUYiSCaCWCeCRCZCVCdCLCCBCkM4K4W4egB4Z4V4d4L4H4P4AEIEEEWQAUKELKI2U2JEFENEDELEUsXEfES8d444KkesBkZkVkdkO4LkHkPkPaA6BqKKCUGkKUUUOUHmZUVUbIdUdk7UEgJMA0I0E0XZAKS0fEm0O0B0bsZ0VIZhUkdk70SUsyQMYMUhesAXRWGMBgOMc0JMFMNMJUQhLMcMMcAwCcH3UgisdkqsFgGsOsWkxsCMFsKRe8RELsbVXsc8AcUXKRCUSUpgT0wsYsPEacXaMWZWJcFcMM%2BWDcLcCM9sDKfcQ8Y8U8OMtk1ya8K4O8Esx8ehV8PAd8UM5sL8KszxJiESbwMSNwYAZMCYPYAcjKAgF0BAe7PSbs1SarIMHERWScv44yJfIkLwTIc0UyIEi2X6UEuyayJQ/OW3edNcgKTc8ybckE/6QGEACjJgk8jc76C8v6ME68nwwIgDOtL4dRaOWOKRWEUc%2BMayL8iQz1IE/OH1LOTUV6UsO0CgoQr0bAIQD4MCa4ZhYbIiJQNBI2JQVvVIC0egQwIhBgR2f0BgNgQQX0csQcscleEop2Dk%2BQVIF5eNVOJ2NSC5IApIOi4UL4ZoU4e0BgZirVF4NiDi7gqCmCgQbmeC74HoWyXsXVWk6ipULabcfUAQNCyCo4RQBqEKHC0yO0OsPI/8iYOjQ4TCWaM/OvOECUmEMISaOWTSpIT80cylXRfRKgNxExFeAAPxcomDcpcQMSMW8qJSSAktdiksRBktUDkttkUtpGUpXH3VEHxCcthMwtgtoHxhCmAHyOiHyn0rws2AIEyGzEUtHLo1cMaObm4oYrkFSGAqEGQUOkY2qqmk1VKK9EwD4oEqaparBzEtASKMCOQUED3kkBkkOB7TAjnPrXPQaI6s2TmsVjAiwEdm5n2zeKSGCOblLQIC4CSCSH/AckTAWl2VkNR12t5lLFfGVLAmMLWqlGFPkoJA4mMMUl5GeogA0FmS4Gq3UkOH/A0E70fyBy%2BoCpYAICj0XTQiQpQryTsWAtmRDREDEGaDRoMGIAWsN0EHclYB6laT0Rmrmon22sMThFoCoFdmAwnn8CsFE38FOBeVEIEiE3znUmMNmQ%2BrZr%2BRYIUAJvoAJSlUkIpqNn1VpqpsEtBuRJojAg%2BFejkCOVi1oAgAoTN22tltQnlsVqOTvxkI1skK1uQuwAAEUuRBbMBZk78UCZbqJtazaLbR4rbDgoD1bP1YbELkLrhEb15kbAlQZMaMbGFmgcaSc8aBbnbN0SbSEybRbIQjYpbN0TB6aUUmaWbPrby3AuatRPDeasUWCYMnbCaCMwVUcxbKD3VJbqa46V5jaUKiQ4Jlb9NVb1a7aUTvbG6Tw1bJcDb265bHbI7CbrbJd8MYagCiQ4bvbfb85/bUaQ7sal8ECI7Lbo70NZrY7QaK6k7aaU6GaD506eRWa2Qs6c7Nl%2BR4xPq%2Ban8gVLbhaRqV4K6Jb3Kpba6L566FalaIAVae715Da677bkLP69be7kDKd%2B6Hbzah76AR729bbUd67B7LbZk3aaiPaJ6p6Eb/KCBvD6Cg7F6idjD8ao7aaY6cQ36h5t7qbk7U7Gbmaj7M67F1Jz69ojBua86BigFc9C76CM1oHVdtrXj464Rn69FX6IHO6m7v6W7f785/737AGG6pH9awHTcJGEIoHkHK6GCx70HB5MGfbsHcGdYOwpt6QCGuGh4iH%2BG16CAN7yGt6E6apqHd7aGD76HDhj72bB9s7BBZkWGXgjAr7LGaceGVEAn4w65u4ERaB8j4ZQahHH6nHRHnHpaEHFHrgu7m64s270mO6lHu6VHjN4HNbFGkHnbYG/dQc3tnl%2BH5G9JF1ujbRlB8x04Qnw7bRVBUciHHh6izEWmUEIAiHVB2HbRHho1VAIBVB1hDhHgIBxnK0mMRb9GvasGkbRyUa24hE8AkEH0l6K7VBbH7H5ry6nH8ZSH17SbO9KGnGd73K9606PGvGs7NlbrdnJlZJ1IBniBFJ/GX4L6gn87JDRsb7eY9RfIwI3m2svnWmf4RnZmEnlmL4n67FN1xG8m5bgHpGcmaj1GsWinJs1HTm4Q8Labp02V1aXQUCMXIGpnZlUhj7pmaWTbzbHh6Xj7xm9GXIDGZ7YZ/am1tndmw6kgDmjmrniWjZzn3KyGTm3ibnKaXH7m3GU6nnGGtF3q/GbqGAdnXs%2BIfmOJvnv8l6QXwai7S0mAIWoWCtDXh14XxnEWH7kXknUXq60nSn8nMmpGf7cn3W5bPXCnQHinwGJX7ZN1yWylKX8NmWNG6X7ZGX1HHa2W43FJOWXjPb4bDH1mJhNnA61pUFhXRWLm7HxXyazmxXN6%2Bms4qGaalX96VWM62aXn/m82hQPn9XIJYWBI/moYAXgBgngWg9n8KkLXpIFAW3YMbWjk7WHWQ2Un0XfXtasXvXcXo38XA3CXs8E3zb4WbbKd43o3E3KnYYqW92U2t3Y2IBUG/792F2WX5nZlL31aOXx6VmM3eXrp%2BW8Hx2C2nHDmi3jmKGzEK6pXibLmK25WgPbnFW9EHm6GG2T6mGNWcHeYx2Mx0A22YWUEjX4CwmglzWIXUPUFMP9Wp3Zl7WuDS2RGXWX6a71H/XsnW6V3b2CmQH28%2B6D3t2j2dHx9Vgb2jaynWWuPyp8Nn2uWrkeWjHvHhAriHAf24Q/3pWwOHGQ2QPpqlPZXAjIOFWa2YPlXD7PG1WOaDXWnu3rJe3%2B3GVXZvnY3U3uUkX5XxbqOxHaPV3dbsXGO/68W3OCWSn%2BP8nHad3JcUCXQmXmPD3tHypgvbOAH/POOIv0A%2BOYuB7BP4vRO02MHVnM2/aNmA6rpGgE5ZP9nf3y3lPKPJWSuNPuDq2aG639PnnEPfHkPJ36VLPcvYYIBrPp3S7EmnWqP140WXPmP6P3PZHYZ6mzF67hufPg2wu4u798MQvz2k35ueOXRovhrB4mnDhvnThRyCuQKL5hnum8bentrrqiHkB0O0BE5K2kgdu9vrihm8b4WemJmpmZm5mFm/lLuNpsRvsGarAuUy6wr7ummG11FQeZPvtClT43vT5FuyPjlgA1ukkLQru/uTVAe7OJ6zEfLIf9vKUsUfvrvBBLhWbsfJVHWrkiH8frjrJie/vN0A05WJOs3kPUaofdumnhXaeHB6f0fE4mfJ8B2RXiv/2S2RfgOKvAPc9qvXHavVXG2GvefkOInAXOGReTWcUVebOEnAsqeacUX%2BvXWZfadEHzauBZkVf%2BffvE49fnLFGl2ZGfWReP7vP13fPzdzfY3d3TdFuH9vflugvKcUeA%2BBOfe7FEvc9A%2BhOEuz2HDhVw/frtvWmueoebeSfoaxOyRWfsvs22vpP9u0/9uefU%2BHu%2BeUABe8babmfgX2nadC3FPi3wPJey3xeW/uCtPE7oO6aFf4OpOB1Gurey%2Bmn%2B398wacUi6deuuH8evZcjeqCTf7fvfLeU%2BUFi%2B6fK/bfBBl%2BMmsmRuXeve9/lGPeZvXfw/Av28FvQvz/Yv734v/9EVQ/BGq0He7/k/ret/M%2BX3uXMv32hQ56NmfEKYzMxuAEQyAXyOgFSJi4mQCgbAInjk7d8dOvfR5v31Pqas1efbIFublw65JMBUTMwDEzibS4Z2EHLOB/UwivR6QqRVopzg0L191CsIYgOcwYEc5nogjNgQ9AX5V0aObrTXm0SRDH9u6y7TzrPwEGTd9%2B03IlhwNpyPQK6oOTdL72zwP5no3xVKsOHB62gDysMI8o8mmbGtOB90eukhVeiO1NCvOfYrXkjgoFTCbeP3C9DWK4E6ByRYWN8BdCi48Ql1GQYwMtz2gVBXOV/qoNOCODGB2g9uJvB/7ic/%2BknYGD8myDwC04YAnoJAOgGMBDYBABISQEQGpMauqAhhkr3VZD8UOrDLARr1a7%2BB76t3CgVQJoEzF6BIvR6LJBYH8DnBYg5wdwLBhL82hj0N3l/REFyN7ejA3oaxz9zsdvBAgp%2BqnEUGj0eO/g%2BgXonUEShNBheZQpvHJT6D4CQwxRiYLMEV4LBbuc/PhlMI347Bx7BwePxcHJE3BHghaONw3wCDfBIoOYcDxWQBDghdQlYYeQiHZ9iQufWejl1iEYpsgu3aWCALaypCGAMGbIXc105998hCHQoTCm%2BamcQA5nbAdw1BZmt/mJQggUQNQDxMKOmncgY70oHUDaBqghoexmYEGDZBXOcYe0OdbG9eBpvGnD0KEEMdRu10O4fIXEHsjOR5UMYS0MegwpG%2BeiK9kZ0pHzCLQQUJYTCjCFhw1heg4VqyIejGCFaZg0vAIMPzH5tC%2BiRYvoXi7UFbBvDdAP7hCH3DnBj0a4WtFuGDCHhW%2BZ4cRlaH9xzRzg%2BURvEFyRCc%2B0QtnpsziECAQR7MZQLJAhFQiiu2nXIXB3hED8J2JnYoYE1KE0jcBoyfAS9DxEEiX%2BBvBRvkw%2BCkjah6xAIZKKYHNCLhdIoUVwMZGL9mRdo5wRIK9bO8mO5YowXyKkGbtuhFY%2BTtMPXiMtHRagmUYcGWHujdBhSb4MqMMGCD8mOw82qWOFjmCq8%2BxUguzHILcxDCMRQQOWFoJmF285wzUZaIejWjPB3IlUfdEeG9jOKQQ10Y9CHHfD0ujSb4jCSOBLkASEJQCFQHwiVQYIYEekNcB1rYBdIZIIhnqDCBPdciC8ZALMmbDNABCp8FYqnSZrCtrI1FQCiACZDasCAAAcX0x5VMAwFKJnvFPjgTXSwAFHNtRYKK0IgYERusEHlpe0hAgHILJCXwBUAOA6wKyBwH8C8A/A3AXgKgE4AnEAe7kTYNsDUi4QeApAX4BwC0BU5SAsoEAP4F%2BpsTJAnEzQDxM4C8BAYv1CSVJNIBwBYAMARACgFtJ3AyAFAYJNGGMkgBgAXAXCHwDoBVhAY39FSaQGiBhA8YnAMSS5OYDMD6Q0QbQFdHcm8Aow5FAgPSAYDZUnJWACEMADcB5FAY3E0gFgD1BGBxAkk3gPgCNC9BqkcUrQEEFiqWxBUqU8gOuLYk5SzoQYZgR4CwBOS9oLAAKaQC1R6YlAu3JKXlUAqpT1gVAAwIvA7AtkRQ9IaeHVP4CCBMa7AKQDIEEA6U1ATk3QJbwMBGAUAmPfQPtEBiwBMYlkrCfQGor1TRgVxBgLKGklNR8QcUsnlxGAymBTilgLgKdUuC0ARyEwMCEBIYA8ThKbFFaSXBqB1BsgLgHENMA6BBAcQQwcoJUD0DpBMg%2BIX6SDKKD4hAZIweIJb26C9AGg8wCGfDNqBXQ%2Bg8wGGcsDhlzABgKM3GS0CxnAyuA6wBQEJJ2ASBWJnADiaQC4k5TeJHAQ4KoASB7xLgE1aapGEODWSXQ/4AcbgEICKgAIJM3gFpJYnrBZJ8k/QJwCUmkBapksumapI4DqSQAmklSesF0kGSSqMcHBqZKjD%2BR6AxACIDak4DMzWZ7M4AJzKkAugxJQoAWSJT0BDThAjCUadIEdmTT1AhUmaaQBFBBhUgAUqmexOUmFSGZVUWcJCEMpMyWZbM4EBbObjczeZEADwOZINkL5/wws8SWrPWAIhlQowd6YpN4ByzfqCs0gAzOVmqz2p4suSQpM4D/gg59MtSRnIrlSyOAZgOuYrJFmZztpMZPwJICAA%3D%3D) - note that `auto` was changed to `int` on lines 218 and 237.

Reply:
I thought the behavior we would want is to see that uint64_t to int is a reasonable type conversion (as long as it doesn't lose bits, which it might depending on int size for the compiler), and that was the purpose of the static_cast you suggested.
Making the type be different wasn't a bug. It was meant to test the tolerance for automatic type promotion.
I don't know if you consider flexible type promotion a virtue or not. That is a question to ask.

Reply:
Is uint64_t to int64_t a reasonable automatic type promotion, or is if its from a const rvalue to an lvalue?

Reply:
Regardless of implicit type promotion, the problem I saw was that there was an error coming from UTransport.cpp's use of Expected when the constructors were combined with a templated constructor. If you think that this is still possible, I will restore that situation and send you the error message I was getting.

Reply:
This is the error message that comes about from trying to combine all the constructors into one templated constructor for the non-Unexpected case:

/home/gzy3g3/projects/up-cpp/src/transport/UTransport.cpp: In member function ‘uprotocol::utils::Expected<uprotocol::utils::callbacks::CalleeHandle<void, const uprotocol::v1::UMessage&>, uprotocol::v1::UStatus> uprotocol::transport::UTransport::registerListener(const uprotocol::v1::UUri&, uprotocol::transport::UTransport::ListenCallback&&, std::optional<uprotocol::v1::UUri>&&)’:
/home/gzy3g3/projects/up-cpp/src/transport/UTransport.cpp:69:33: error: could not convert ‘std::move<uprotocol::utils::callbacks::CalleeHandle<void, const uprotocol::v1::UMessage&>&>((* & handle))’ from ‘std::remove_reference<uprotocol::utils::callbacks::CalleeHandle<void, const uprotocol::v1::UMessage&>&>::type’ {aka ‘uprotocol::utils::callbacks::CalleeHandle<void, const uprotocol::v1::UMessage&>’} to ‘uprotocol::utils::Expected<uprotocol::utils::callbacks::CalleeHandle<void, const uprotocol::v1::UMessage&>, uprotocol::v1::UStatus>’
   69 |                 return std::move(handle);
      |                        ~~~~~~~~~^~~~~~~~
      |                                 |
      |                                 std::remove_reference<uprotocol::utils::callbacks::CalleeHandle<void, const uprotocol::v1::UMessage&>&>::type {aka uprotocol::utils::callbacks::CalleeHandle<void, const uprotocol::v1::UMessage&>}

Reply:
The templated constructor that I'm trying to merge the prior indivdual constructors into is this:

	template <class X = T>
	constexpr explicit Expected(X&& rhs) {
		std::get<T>(storage_) = static_cast<T>(std::forward<X>(rhs));
	}

Reply:
If you can recommend a fix for this, I'm happy to try it.

Reply:
This pattern of multiple constructors and trying to debug stuff on the constructor signature could probably be solved in a more general, and much more readable way with one templated constructor and 'if constexpr ()' inside the constructor. Is that acceptable as a path forward?